### PR TITLE
Implement satisfiers

### PR DESCRIPTION
**Pull request by wei2912**
I'll do a short write up when I finish the test suites.

### CODE REVIEW

**Comment by wei2912** [comment]
@bnjmnt4n My apologies in advance for the size of this PR lmao

Tldr is that I implemented a `Satisfier` type in `satisfiers/types.ts`, which aids processing of modules using a tree-like approach. Logic can be specified in each satisfier to determine how modules are assigned/filtered, or to enforce some constraint on modules.

There are two types of satisfiers:

1. `SatisfierBranch`, which provides a list of satisfiers and optionally a `reduce` function that will use the results from the satisfiers to determine if some constraint is satisfied.
2. `SatisfierLeaf*`, which processes `assigned`/`remaining` modules, or enforces some constraint.

The magic happens in `satisfiers/index.ts` which will recursively evaluate the satisfiers when given a list of `assigned`/`remaining` modules. By handling just satisfier objects, this allows satisfier logic to be handled in a consistent and sane manner, e.g. these examples from `matchRule/satisfiers.ts`:

```ts
const patternMatchRuleSatisfier = (
  ref: string,
  rule: PatternMatchRule
): SatisfierLeafAssign => {
  const { pattern, exclude = [], info } = rule;
  const excludeArray = typeof exclude === "string" ? [exclude] : exclude;
  const excludeRE = patternToRE(...excludeArray);
  const { assign } = patternSatisfier(ref, pattern);
  return {
    ref,
    assign: (remaining) => {
      const { remainingMask } = assign(remaining);
      return {
        remainingMask: remainingMask.map((bool, i) => {
          const module = remaining[i];
          const [moduleStr] = module;
          return bool && !excludeRE.test(moduleStr);
        }),
      };
    },
    info,
  };
};

const andMatchRuleSatisfier = (
  ref: string,
  { and }: AndMatchRule
): SatisfierBranch => ({
  ref,
  satisfiers: and.map((rule, i) =>
    matchRuleSatisfier([ref, i].join("/"), rule)
  ),
  reduce: (results) => ({
    isSatisfied: results.every(({ isSatisfied }) => isSatisfied),
  }),
  message: "modules do not match all rules",
});
```

A `ref` is used to keep track of where the original rule was invoked. Also, some black magic is done to create a sort of new "context", e.g. a rule that requires a block to be satisfied would pass all the `assigned` modules to that block as `remaining` modules to be processed.

**Comment by wei2912** [comment]
To test the code, you can refer to `src/index.ts` and do something like this:

```ts
const { primary } = initDirectories();
const modules: Module[] = ...; // see module/types.ts for how to define an array of modules
const blockId = "cs-hons-2020"; // you can choose the filename of any block in `blocks/primary`, excluding the file extension
const result = verifyPlan(modules, primary, blockId);
```

**Comment by wei2912** [comment]
Merging in, we can write a more detailed test suite later.
